### PR TITLE
Optionally create missing Flags, Samples and Switches

### PR DESCRIPTION
**Pull request by adamhaney**
Flags, Switches and Samples may be scattered throughout templates,
backend code or across mulitple frontend clients. This patch provides
a user with the option of having waffle create a missing Flag, Sample
or Switch when one is requested but does not exist.

The goal of this change is to save operators time by removing the need
to search through code for flag/switch/sample names and to reduce
communication requirements accross teams with of multiple developers
with multiple environments that might require the creation of a
flag/switch/sample in the database of every environment code is run
in.

# Behavior Changes
## Old Behavior
  1. Flag/Switch/Sample is requested
  2a. If found in db evaulate what the value should be for request
  2b. If missing return (FLAG|SWITCH|SAMPLE)_DEFAULT value
## New Behavior
  1. Flag/Switch/Sample is requested
  2a. If found in db evaulate what the value should be for request
  2b. Create Flag/Switch/Sample with
      value (FLAG|SWITCH|SAMPLE)_DEFAULT
  3. Return (FLAG|SWITCH|SAMPLE)_DEFAULT for this request, subsequent
  requests will be served using the value from the cache/db

# Further Discussion
  - #44 (https://github.com/jsocol/django-waffle/issues/44)
  - #50 (https://github.com/jsocol/django-waffle/issues/50)

# Future Work
This change intentionally narrowly scoped and as such does not provide
a mechanism to set a default value other
than (FLAG|SWITCH|SAMPLE)_DEFAULT for individual objects that are not
in the database although it could be expanded in the future to do so.

### CODE REVIEW

File: waffle/models.py
Comment (clintonb) [inline waffle/models.py]: What would cause these errors? If such an error occurs, I would rather see it bubble up that be squashed with an uninformative log line.

File: waffle/models.py
Comment (clintonb) [inline waffle/models.py]: You want to use the setting value. This can be combined with the line above: `get_setting('SAMPLE_DEFAULT', 0)`.

File: waffle/models.py
Comment (clintonb) [inline waffle/models.py]: Why are you creating a transaction for all of these?

**Review by clintonb — CHANGES_REQUESTED** [review]
I must admit that I am not a fan of implicitly creating data in this manner. You run the risk of creating a bunch of data that needs to later be cleaned up. Why not use a migration to create the flag/sample/switch?

File: waffle/models.py
Comment (adamhaney) [reply waffle/models.py]: I wasn't trying to just get a default value, or at least I don't think I was :)

In the defaults `SAMPLE_DEFAULT` is a boolean value, however there isn't a boolean option for a Sample, it's expressed as a percentage, so I'm saying if `SAMPLE_DEFAULT` is true then that implies that you want the sample value to be 100, if I just grabbed the default value I'm testing if it's defined.

I've never really used Samples in production so I may be misunderstanding their intended use.

File: waffle/models.py
Comment (adamhaney) [reply waffle/models.py]: The IntegrityError would be thrown if you attempt to create a flag name that already exists, which can happen if two processes both encounter a flag that hasn't been created and attempt to create it. I caught it because I my philosophy around this change set was very "first do no harm" since it's intended to be a convenience and the behavior of returning the FLAG_DEFAULT is unaffected by whether or not the flag could be created I thought it was best to catch it.

Still, you're the maintainer so if you don't like that explanation I can definitely change it.

File: waffle/models.py
Comment (adamhaney) [reply waffle/models.py]: In Postgres (not sure about other databases) if a cursor throws an IntegrityError then later interactions with the database will fail (even if you catch the exception) [1]. So I'm just trying to make sure that failure to create flags doesn't take down what would otherwise be a perfectly healthy request.

1. https://docs.djangoproject.com/en/2.1/topics/db/transactions/#controlling-transactions-explicitly

Although, upon further thought I could probably accomplish a similar thing by using `get_or_create` instead of just `create`.

File: waffle/models.py
Comment (clintonb-stripe) [reply waffle/models.py]: Got it. I've never used samples, either.

Comment (clintonb) [reply waffle/models.py]: Got it. I've never used samples, either.

**Comment by adamhaney** [comment]
> I must admit that I am not a fan of implicitly creating data in this manner. You run the risk of creating a bunch of data that needs to later be cleaned up. Why not use a migration to create the flag/sample/switch?

I'm currently working on a project where the goal is to wrap every new feature in a flag, but sometimes things fall through the cracks and code ends up in the codebase that people don't realize is there, or that is turned off behind a flag that people forget the name of. Ironically because it's so easy to drop flags into templates, and python, and even javascript it's hard to get a top level view of every flag that's being evaluated. 

Also, in an effort to use descriptive flags that don't have name collisions we use names like, "jira_32_send_users_weekly_email" which is really easy to typo and then you end up with a subtly typo'd flag name in the admin that you *think* is turning a flag on or off but instead it's just quietly evaluating the default.

I completely acknowledge the risk of accidentally implicitly creating data, to maintain continuity with the way this library has been operating for the last couple of years I wouldn't suggest making this setting on by default. In general it would only impact users of the library that are explicitly asking to have data created for them.

**Comment by adamhaney** [comment]
Also, if you're considering accepting this patch I did realize I need to write some regression tests before it's merged.

**Comment by clintonb** [comment]
I would solve the problem with tests and/or a checklist to ensure flags are created as part of the development process. This would solve the issue of typos and forgetting flags for server-side code.

That said, I understand the desire to do things this way. Ping me once you have tests, and I'll review again.

**Comment by adamhaney** [comment]
@clintonb tests are added, I covered testing creation with the setting enabled and positive and negative paths for defaults please let me know if there are any additional tests you'd like to see, and thanks for taking the time to review.

Also FWIW I deployed my fork to our production service this morning and haven't seen any problems (although that's obviously anecdotal).

File: waffle/models.py
Comment (clintonb) [reply waffle/models.py]: Log as a warning, but include the stack trace by passing `exc_info=True` in your call to `warning`. Do that in all of the new exception handlers

File: waffle/models.py
Comment (clintonb) [inline waffle/models.py]: `default_percent = 100 if get_setting('SAMPLE_DEFAULT') else 0`

File: waffle/tests/test_waffle.py
Comment (clintonb) [inline waffle/tests/test_waffle.py]: Combine into a single call: 
```python
FLAG_NAME = 'my_dynamically_created_flag'
flag = flag_model.objects.get(name=FLAG_NAME)```

File: waffle/tests/test_waffle.py
Comment (clintonb) [inline waffle/tests/test_waffle.py]: Extract all of this into a helper method (e.g. `assert_flag_dynamically_created(expected_value)`).

File: waffle/tests/test_waffle.py
Comment (clintonb) [inline waffle/tests/test_waffle.py]: Same notes as above: create a helper method, load the object by name.

File: waffle/tests/test_waffle.py
Comment (clintonb) [inline waffle/tests/test_waffle.py]: Same notes as above: create a helper method, load the object by name.

**Review by clintonb — CHANGES_REQUESTED** [review]

**Comment by adamhaney** [comment]
@clintonb I took forever to get back to this change set but I think I've addressed all your requested changes.

**Review by clintonb — APPROVED** [review]

**Comment by clintonb** [comment]
@adamhaney would you add documentation, please?

**Comment by adamhaney** [comment]
@clintonb documented all of the settings in the "Configuring Waffle" section and added a section about the optional auto creation behavior on each type's page. I'm Happy to make any revisions based on feedback.

File: docs/starting/configuring.rst
Comment (clintonb) [inline docs/starting/configuring.rst]:
```suggestion
    If Waffle encounters a reference to a flag that is not in the database, should Waffle create the flag?
```

File: docs/starting/configuring.rst
Comment (clintonb) [inline docs/starting/configuring.rst]:
```suggestion
    If Waffle encounters a reference to a switch that is not in the database, should Waffle create the sample?
```

File: docs/starting/configuring.rst
Comment (clintonb) [inline docs/starting/configuring.rst]:
```suggestion
    If Waffle encounters a reference to a sample that is not in the database, should Waffle create the sample?
```

File: docs/types/sample.rst
Comment (clintonb) [inline docs/types/sample.rst]:
```suggestion
always evaluates to ``False``.
```

File: docs/types/flag.rst
Comment (clintonb) [inline docs/types/flag.rst]:
```suggestion
auto-created database record.
```

File: docs/types/switch.rst
Comment (clintonb) [inline docs/types/switch.rst]:
```suggestion
missing switch you can set :ref:`WAFFLE_CREATE_MISSING_SWITCHES
```

File: docs/types/switch.rst
Comment (clintonb) [inline docs/types/switch.rst]:
```suggestion
<starting-configuring>` to ``True``. Missing switches will be created in
```

File: docs/types/switch.rst
Comment (clintonb) [inline docs/types/switch.rst]:
```suggestion
auto-created database record.
```

**Review by clintonb — CHANGES_REQUESTED** [review]

**Comment by guybowden** [comment]
I came here looking to see if anyone had created a set of simple migration commands (i.e. `manage.py makewaffleflagmigration SOME_FLAG_NAME`) to generate the instances - this looks like it'll negate the need for this. Good work guys.

**Review by clintonb — APPROVED** [review]

**Pull request merged**
